src/formatter/javascript.ts

```diff
@@ -211,18 +242,76 @@ export class SyntheticsGenerator extends JavaScriptLanguageGenerator {
     return `});`;
   }
 
-  generateText(actions: Array<ActionInContext>) {
+  /**
+   * Generates JavaScript code from a custom set of steps and nested actions.
+   *
+   * This function makes no assumptions about where steps should be created,
+   * and instead follows the step definitions the caller has defined.
+   * @param steps IR to use for code generation
+   * @returns a list of the code strings outputted by the generator
+   */
+  generateFromSteps(steps: Steps, filterEmptyLines = true): string[] {
     const text = [];
     if (this.isSuite) {
       text.push(this.generateHeader());
     }
-    for (let i = 0; i < actions.length; i++) {
-      text.push(this.generateAction(actions[i]));
-      if (i === actions.length - 1) text.push(this.generateStepEnd());
+    for (const step of steps) {
+      if (step.actions.length === 0)
+        throw Error('Cannot process an empty step');
+      text.push(
+        this.generateStepStart(
+          step.name ??
+            step.actions[0].title ??
+            actionTitle(step.actions[0].action)
+        )
+      );
+
+      for (const action of step.actions) {
+        text.push(this.generateAction(action, true, true));
+      }
+
+      text.push(this.generateStepEnd());
     }
     if (this.isSuite) {
       text.push(this.generateFooter());
     }
+    return text.filter(s => !!s || !filterEmptyLines);
+  }
+
+  /**
+   * Generates JavaScript code for the given list of actions.
+   * @param actions The actions to convert to code.
+   * @param overrideStepOrder When `true`, all passed actions will be nested in one step.
```

Comment:
Similar concerns to the comment I left for `ignoreStepDefaults`. Do you think something like `wrapIntoStep` could work?